### PR TITLE
grouping prototype

### PR DESCRIPTION
**Pull request by piever**
This is a WIP to play around with ideas from https://github.com/JuliaPlots/Plots.jl/issues/1530 in StatMakie.

Instead of a `group` keyword I'm using the first argument as I don't know how to implement this in Makie otherwise...

The idea is that one can use the first argument to do some grouping and define some keywords on those grouping. Here for example I'm plotting a scatter plot of `rand(10)` versus `rand(10)` using two vectors `rand(Bool,10)` and `rand(Bool, 10)` to specify color and markersize respectively. Then I'm passing the list of colors and markersizes (this part should be optional and default to something sensible, but it is not at the moment).

```julia
julia> scatter((:color => rand(Bool, 10) => [:blue, :red], :markersize => rand(Bool, 10) => [0.03, 0.2]), rand(10), rand(10))
```

screenshot from 2018-10-03 15-33-12

Would be happy to have feedback on syntax (I'm starting to like the `:color => v => scale` syntax but different people may have different taste) and on whether this should be closer to Plots grouping (meaning only work for categorical values) or to GoG aesthetics (also accept continuous values and try to detect automatically which it is).

In combination with `@df` this could be used with column names rather than vectors and should provide some automated legend entries for the group as well.

### CODE REVIEW

**Comment by piever** [comment]
At the moment one can also pass a function or a dict rather than an array as a scale, for example:

```julia
julia> scatter((:color => rand(Bool, 10) => t -> t ? :blue : :red, :markersize => rand(Bool, 10) => [0.03, 0.2]), rand(10), rand(10))
```

**Comment by piever** [comment]
This now requires https://github.com/JuliaPlots/AbstractPlotting.jl/pull/35 and https://github.com/JuliaPlots/Makie.jl/pull/203 if people want to try it out.

I now use a special `Group` type that accepts keyword arguments:

```julia
N = 20

scatter(
    Group(
        color = rand(Bool, 10) => [:blue, :red],
        markersize = rand(Bool, 10) => [0.03, 0.2]),
    rand(10),
    rand(10)
)
```

and I've started adding default scales:

```julia
scatter(
    Group(color = rand(1:4, N), marker = rand(Bool, N)),
    rand(N),
    rand(N)
)
```

screenshot from 2018-11-03 16-59-17

In this case `marker` cycles across all possible markers (now it's a bit hacky as they are only accessible as a `Dict` from `AbstractPlotting` so the order is undefined, but this can be easily fixed from there) and `color` default to the beautiful `to_colormap(:Dark2)`  of https://github.com/JuliaPlots/AbstractPlotting.jl/pull/37

Important design remark: Makie is quite different from Plots in that there is no pipeline to process keywords sequentially, so for `group` to have such a drastic effect on the plot, it needs to be a positional argument so the API will probably stay as is.

**Comment by piever** [comment]
I feel it's making good progress. Now I can have some GoG flavor already.

Default grouping is by color, like in Plots. Positional arguments and "styling that does not group" goes into Style. Outside Style are arguments that don't use the dataset.

```julia
using StatsMakie, RDatasets
julia> iris = RDatasets.dataset("datasets", "iris");

julia> scatter(iris, Group(:Species), Style(:SepalLength, :SepalWidth))
```

screenshot from 2018-11-04 13-11-43

## Grouping by more than one thing

```julia
julia> mpg = RDatasets.dataset("ggplot2", "mpg");

julia> scatter(mpg,
           Group(marker = :Manufacturer, color = :Class),
           Style(:Displ, :Hwy), markersize = 1
       )
```

screenshot from 2018-11-04 13-11-19

# Give a pair of column and scale to use a custom scale:

```julia
julia> scatter(mpg,
           Group(marker = :Manufacturer, color = :Class => [:blue, :cyan, :magenta]), #cycle through these colors
           Style(:Displ, :Hwy), markersize = 1
       )
```

screenshot from 2018-11-04 13-32-53


## Adding a continuous color

```julia
julia> scatter(mpg,
           Group(marker = :Manufacturer),
           Style(:Displ, :Hwy, color = :Hwy), markersize = 1 
       )
```

screenshot from 2018-11-04 13-29-54

## Building complex graphics from simple ones

The user can pass many group objects and styles that are merged together, for example:

```julia
julia> scatter(mpg,
           Group(marker = :Manufacturer), Group(:Class),
           Style(:Displ, :Hwy), markersize = 1 
       )

julia> scatter(mpg,
           Group(marker = :Manufacturer),
           Style(:Displ, :Hwy), Style(color = :Hwy),  markersize = 1 
       )
```

which is the equivalent of adding things with the `+` in ggplot2. I think it makes it easier to reuse "building blocks" across plots, without overloading `+` which would feel a bit strange.

## TODO list:

- [ ] Add legend / labelling (not sure how they work in Makie)

- [ ] Simplify implementation when new recipe system gets finalized in Makie

- [ ] Add "facets" (not sure how to do though as subplotting is not really implemented via keywords here, unlike the other "aesthetics")

- [ ] Fix some `markersize` bugs in AbstractPlotting I think: 

```julia
julia> scatter(Group(marker = rand(Bool, 10)), rand(10), markersize = 0.1:0.1:1)

Error showing value of type Scene:
ERROR: MethodError: no method matching glyph_scale!(::Observables.Observable{Char}, ::Array{Vec{2,Float32},1})
Closest candidates are:
  glyph_scale!(::Char, ::Any) at /home/pietro/.julia/dev/AbstractPlotting/src/utilities/texture_atlas.jl:128
```

**Comment by Evizero** [comment]
Looking good! The one comment I have is that to me it seems more general to have `group1 + group2` denote merging groups, instead of specially supporting multiple group arguments in some functions.

**Comment by piever** [comment]
That's a good point. The support for many groups should happen for all recipes if I understand the Makie pipeline correctly (I'll be honest, that's a big if :) ). OTOH I agree that it's generally useful to play with this group / style objects as a user outside of plot calls: I guess one may even want to save them to disk for future sessions.

Now you can manually combine groups and styles with each other using `merge` (say `merge(grp1, grp2)`) and indeed the pipeline simply does `foldl(merge, groups)`. OTOH maybe it'd be nicer to have some operator as well rather than `merge`.  `+` felt a bit strange - we don't even use it for string concatenation in Julia so it really is a bit out of nowhere -, but I am unsure what other options are there. I was also considering `*` (from stirng concatenation) and `∪` (from union), but I'm not thrilled about those either... After all, given that you explained "it'd be nice to have `+` to merge groups", maybe `merge` is indeed the right choice :)

**Comment by Evizero** [comment]
Right, I actually edited my post as well about `+` vs `*`. 

I think as a first step it might make sense to just go with `merge`. This way we don't promise convenience API details too early. As time goes on we can always overload `*` if we feel like writing `merge` is such a hassle.

**Comment by piever** [comment]
I like the argument for `*` as merging is associative but not commutative (rightmost one wins): I think I'll go for it.

**Comment by mkborregaard** [comment]
This is really progressing sweetly. Question: would it be clearer to do `Group(color = :Class), color = [:blue, :cyan, :magenta])` which seems more native than `color = :Class => [:blue, :cyan, :magenta]`. Or - does it actually support this already?

**Comment by mkborregaard** [comment]
> I like the argument for * as merging is associative but not commutative (rightmost one wins): I think I'll go for it.

FWIW I had the thought that `*` could be used for merging Makie `Theme`s as well, for the exact same reason that it's not commutative (same argument as using it for Strings essentially. Like I think it would be cool to have the concept of style themes and color themes as separate things that the user might merge with `*`.

**Comment by piever** [comment]
> Question: would it be clearer to do `Group(color = :Class), color = [:blue, :cyan, :magenta])` which seems more native than `Group(color = :Class => [:blue, :cyan, :magenta])`?

We mentioned this option already at https://github.com/JuliaPlots/Plots.jl/issues/1530 and I initially decided against it here, but looking more closely it's a very nice solution.

My "pair" strategy never seemed to impress instead the `color` keyword is unused if the user is setting colors via a group,  and your proposed solution plays very nicely with themes.

The only thing I'd like to mention is that we may need to allow using custom types for fancy scale behavior. For example, let's say I have `"Placebo", "Treatment 1", "Treatment 2"` and I want them to always be `"black", "red", "blue"` respectively for my paper. Maybe some plots only compare `"Placebo"` and `"Treatment 2"`: the palette would make `"Treatment 2"` red instead of blue, which I don't want. So instead of passing a normal `Palette` we would need some custom type that encodes the correct mapping.

**Comment by mkborregaard** [comment]
Good point, but I think quite often that would mostly happen 1) in connection to facets or some other behaviour that groups over the entire dataset to create subplots/series, in which case the problem would not arise. In the relatively rare case where you'd want to create a single plot just on a subset of the data but still have the colors comparable to other unrelated plots (/scenes) in the paper, would that not be easily taken care off by passing a color vector explicitly? `color = mypalette.colors[1:3;5]`

**Comment by piever** [comment]
So, I've reimplemented it the way we suggested. What I'm doing is, I'm creating a trait `isscale` that says if the value of an attribute corresponds to a `scale` (I think arrays, functions and dicts all make sense for example). The attributes are generally non-empty because AbstractPlotting decides to put things in them, so there needs to be a criterion to know if it's a proper value or a placehodler.

**Comment by piever** [comment]
# Statistics proposal

There was some discussion on slack that I would like to continue here. I'm writing in this issue rather than opening a new one because the discussion is much more related to the Grammar of Graphics story than I thought.

Some "recipes" that we normally use (and some that we haven't implemented yet) are primarily computing some statistics rather than changing the visualization. For example `density(v)` is computing the kernel density estimate of `v` pdf (and plotting as a line). If `v = randn(1000, 2)` for example, then `density(v)` could still compute the `2D` kernel density and plot as `Heatmap` by default (but `Surface` and `Contour`) also makes sense. Same thing applies to several other statistics (`Histogram`, `LinearRegression`, `LoessRegression`, `Count`). In Grammar of Graphics terms those are called Statistics and each of them has a default Geometry (i.e. a `PlotType`).

## How to incorporate Statistics and Geometry in StatsMakie

One may think that if you want to plot the statistics, you should just compute it and plot the result, i.e.

```julia
plot(kde(v))
surface(kde(v))
```

which is a reasonable "purist" point of view. However it doesn't play nice with grouping in that now we can no longer use the `Group` technique, as we should group before running the statistics.

### Adding a statistics type

I think that here we do not need more recipes but simply to add the concept of a `Statistics` (or `Model`), where the first argument can be given as a `Statistics` (say `Histogram`, `Count`, `Density`, `Loess`) and (through a mechanism very similar to `convert_arguments`) StatsMakie would put together `Group`, `Style` and `Statistics` and execute them in the correct order (first extract columns, then group, then fit the statistics on the groups, taking away from the plot the keywords that were destined to the statistics).

For example

```julia
plot(Histogram, randn(100), nbins = 30)
surface(Density, randn(1000, 2), bandwidth = (0.1, 0.2))

scatterlines(df, Density, Group(:Species), Style(:SepalLength))

lines(df, Loess, Group(:Species), Style(:SepalLength, :SepalWidth))
```

As a technical point, this has the nice advantage that we can expect a `Statistics` to now how to deal with labels (meaning, `Histogram` + `Style(:SepalLength)` would return as labels `("SepalLength", "Count")`.

On top of that (which has to be considered as a `convert_arguments` method, with the twist that it consumes some keywords) one can add some short-hands

```julia
density(v) = plot(Density, v)
```

which importantly are not recipes, meaning that `density(v)` will automatically get the correct theme for whatever `plottype(Density, v)` is.

## Bikeshedding

One thing I'm unsure about is how to combine these things with an operator. We had decided on `*` to join `Group` but while that makes sense as joining `Groups` is non-commutative (in case of repeated attributes), joining a `Group` with a `Statistics` is commutative, so maybe we could keep the `*` just to join things of the same type (`Group` with `Group`, `Style` with `Style`).

**Comment by mkborregaard** [comment]
I definitely think Makie should support direct function plotting - doesn't it already? :-O

**Comment by SimonDanisch** [comment]
> doesn't it already? :-O

I think so ;) That's actually a cool idea - we could just use the actual statistics function as an argument, and make it work the same way as if you'd do `lines(0..1, 0..1, func)` :) Should also work with grouping, I suppose

**Comment by piever** [comment]
The interesting thing, if passing a function, is that this would be (both conceptually and most likely in the implementation) the "split apply combine" strategy, in that we split by `Group`, apply our function and combine in the plot. So:

```julia
plot(iris, Group(:Species), Style(:SepalLength)) do v
    fit(Histogram, v, nbins = 30)
end
```

which we could support in general. It seems like this whole concept of Statistics can be replaced by a function and some shorthands..., I'll try to code in a prototype here

**Comment by mkborregaard** [comment]
sweet 💖

**Comment by piever** [comment]
I think this is now a in-principle complete implementation of a Julia version of Grammar of Graphics (meaning, it still lacks a lot of features and it needs some recipes on the Makie side to accomodate legends and layout, but all the building blocks are there.

To see an example with all building blocks:

```julia
 plot(kde, #statistics
   iris, #data 
   Group(linestyle = :Species), #categorical grouping
   Style(:SepalWidth), #setting aesthetics (here I may probably also accept symbols directly)
   linewidth = 2, # keywords for the plot
   bandwidth = 0.2 #keywords for the statistics (smoothing)
)
```

screenshot from 2018-11-12 13-51-55

The support for Geometry is in the plotting function: `plot` gives the default geometry but one could use for example `scatter(kde, ...)` (even though in this particular case it makes limited sense).

In terms of implementation, this relies heavily on the [new and improved convert_arguments](https://github.com/JuliaPlots/AbstractPlotting.jl/pull/42) and a recipe for multiple series that I've hacked together here, but hopefully will be replaced by a more complete one in AbstractPlotting that also takes care of layout and legends and labels.

This needs a little bit more clean up and tests and then I think is ready to go (provided https://github.com/JuliaPlots/AbstractPlotting.jl/pull/42 gets merged with the fix for kwargs in [this comment](https://github.com/JuliaPlots/AbstractPlotting.jl/pull/42#pullrequestreview-173669246)).

**Comment by SimonDanisch** [comment]
Awesome :) I just pushed the discussed fix! If it all works for you, feel free to merge JuliaPlots/AbstractPlotting.jl#42 !

**Comment by Evizero** [comment]
a minor comment: In case of plotting some function that is provided via a parameter, I think its slightly more useful to label the y-axis `f(x)` instead of `y`.

offtopic: i keep hoping that one day the awesome makie-screenshots that you are posting won't feature the giant bold and italic ticks and axis labels :-)

File: src/tables.jl [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+end
+
+Base.merge(s1::Style, s2::Style) = Style(s1.args..., s2.args...; merge(s1.kwargs, s2.kwargs)...)
+Base.:*(s1::Style, s2::Style) = merge(s1, s2)
+
+Base.merge(g::Group, s::Style) = merge(Style(g), s)
+Base.merge(f::Function, s::Style) = merge(Group(f), s)
+Base.merge(s::Style, g::Union{Group, Function}) = merge(g, s)
+
+const GroupStyle = Union{Style, Group}
```
Comment (Evizero) [inline src/tables.jl]: maybe `GroupOrStyle`? similar to `VecOrMat` in Base

**Comment by piever** [comment]
> a minor comment: In case of plotting some function that is provided via a parameter, I think its slightly more useful to label the y-axis f(x) instead of y.

Labelling should definitely be improved but there is still no framework to do so. One idea is that each "BasicPlot" here would give their own "recommendation" of a possible label and a scene recipe will take care of choosing the most appropriate option. Hopefully in most cases that option will be preferable to a gigantic `x, y`.

> offtopic: i keep hoping that one day the awesome makie-screenshots that you are posting won't feature the giant bold and italic ticks and axis labels :-)

I hope so too, but I haven't been very proactive about it...

File: src/tables.jl [hunk 2]
```diff
@@ -0,0 +1,54 @@
+struct Style
+    args::Tuple
+    kwargs::NamedTuple
+    Style(args...; kwargs...) = new(args, values(kwargs))
+end
+
+Base.merge(s1::Style, s2::Style) = Style(s1.args..., s2.args...; merge(s1.kwargs, s2.kwargs)...)
+Base.:*(s1::Style, s2::Style) = merge(s1, s2)
+
+Base.merge(g::Group, s::Style) = merge(Style(g), s)
```
Comment (Evizero) [inline src/tables.jl]: how does `Style(g)` work when `g::Group` ? what method gets called?

**Comment by Evizero** [comment]
> I hope so too, but I haven't been very proactive about it.

It was just a cheeky comment. I hope you know that I really do **not** wish to distract you from the far more valuable work that you are doing

**Comment by piever** [comment]
This is basically ready, I think it only needs some bikeshedding on names, argument order, etcetera.

Here we introduce 3 things: grouping support, "statistic" support and table support.

Grouping is done by passing a `Group` object as first argument:

```julia
scatter(Group(marker = rand(Bool, 100)), rand(100), rand(100))
```

"Statistics" (Grammar of Graphics terminology) is done by passing a function as first argument:

```julia
plot(kde, rand(100))
```

They can be combined:

```julia
plot(kde, Group(linestyle = rand(Bool, 100)), rand(100))
```

Table support is done by passing the table as first argument and the columns can be given as `Style`, regardless of whether they are a positional of keyword argument:

```julia
scatter(iris, Style(:SepalLength, :SepalWidth, color = :PetalLength))
```

It can also be mixed with a "statistic" and I use the same order as IndexedTables `groupby(f, data, by; select)`:

```julia
plot(kde, iris, Group(linestyle = :Species), Style(:SepalLength))
```

Here the order of `Style` and `Group` can actually be changed.

I'm reasonably happy with the API (we can add extra methods to simplify some scenarios later), but one remark was that while `Style` makes a lot of sense for attributes, say `Style(color = :SepalLength, markersize = :SepalWidth)`, it works less well for positional arguments, say `scatter(iris, Style(:SepalLength, :SepalWidth))` (even though one can argue we are styling the x and y coordinates).

OTOH, it is quite nice to have a unique concept for positional arguments and attributes, so I was planning to merge as is unless there are strong counter proposals wrt API.

**Comment by piever** [comment]
Another issue that it'd be nice to figure out is the following: `kde` and `fit!(Histogram, ...)` etc. tend to take arguments as a `Tuple` in the multidimensional case. Say:

```julia
kde((rand(100), rand(100)) # for a 2D distribution
```

This leads to:

```julia
plot(kde, Style((:x, :y))
```

Which looks a bit clumsy. Should we have our own `density` that is a bit simpler and allows `density(x, y) = kde((x,y))`?

**Comment by mkborregaard** [comment]
I think it makes sense to implement local functions for `density` etc that are tuned to giving a nice plotting api :+1:
To me, `Style` is a confusing term for positional arguments, but it's really the same as `aes`thetic in ggplot2, so I might be the only one who feels this. I do think I prefer it over `plot(kde, iris, :SepalLength, Group(linestyle = :Species))` which has a bit too many positional args to remember the position of.

Finally I'll say it's really impressive how you've managed to implement a nice and understandable Julia-ish API on Makie that allows very powerful GOG. This is one of my favourite PRs I've seen in a long time.

**Comment by Evizero** [comment]
in this example:

```julia
scatter(Group(marker = rand(Bool, 100)), rand(100), rand(100))
```

how do i specify the two markers to choose from? My first naive guess based on earlier examples would be

```julia
scatter(
    Group(marker = rand(Bool, 100)),
    rand(100), rand(100),
    marker = [:circle, :rect]
)
```

right? and if yes is this also true for tables?

**Comment by piever** [comment]
> how do i specify the two markers to choose from?

Yes, you guessed correctly, it's:

```julia
scatter(
    Group(marker = rand(Bool, 100)),
    rand(100), rand(100),
    marker = [:circle, :rect]
)
```

and it also works for tables. The nice thing about wrapping things in `Style` is that the `marker` attribute won't be inside `Style` so Julia will not confuse its symbols with potential names of columns (as instead could happen in `@df` in StatPlots).

A cleaner way (proposed by Michael), is to set this thing in the theme:

```julia
julia> new_theme = Theme(
           scatter = Theme(
               marker = [:rect, :circle], # don't do this! it works in this case, but we should use a Palette type!
               markersize = 0.3
           )
       )

julia> set_theme!(new_theme)

julia> scatter(
           Group(marker = rand(Bool, 100)),
           rand(100), rand(100)
       )
```

Even though for colors we have the nice `Palette` that can serve both as a list and as a value for a specific plot, whereas here passing directly a vector in the theme can cause issues. So the next step is to implement a generic `Palette` type to be able to set all these default values in the theme.

**Comment by dpsanders** [comment]
This looks very nice!

To me it would be more natural to do something like
```
plot(KDE(bandwidth = 0.1), rand(100), color = :blue)
```
It at first seems kind of cool to be able to mix the different types of keyword arguments, but in the end I think it is more confusing to do so.

**Comment by mkborregaard** [comment]
@dpsanders yes, but how would you do that in practice? Given that `plot` isn't a macro, how would you use `KDE(;kwargs...)` when that method doesn't exist? You could do  `plot(x->KDE(x, bandwidth = 0.1), rand(100), color = :blue)` already, I think.

@piever Could the `Theme` constructor not simply wrap an input Vector in a `Palette` without the user having to worry about this?

**Comment by dpsanders** [comment]
Define a `KDE` object and dispatch on it?

**Comment by mkborregaard** [comment]
Yes but that's already possible, no? That's just defining recipes, like there should also be a recipe for `StatsBase.Histogram` independent of what other calls can be used to call a `histogram`.

**Comment by mkborregaard** [comment]
Ah, I see what you mean - then you wouldn't get the extraction. So it would be to call
```julia
plot(KDE(bandwidth = 0.1), iris, Style(:SepalWidth))
```
?

**Comment by dpsanders** [comment]
Yes, exactly.

**Comment by piever** [comment]
> To me it would be more natural to do something like
> ```julia
> plot(KDE(bandwidth = 0.1), rand(100), color = :blue)
> ```

That's a very good point. However would currying address thins?

Meaning, it'd be enough to add a method:

```julia
kde(; kwargs...) = (args...) -> kde(args...; kwargs...)
```

and then

```julia
plot(kde(bandwidth = 0.1), iris, Style(:SepalWidth))
```

would work as expected. If the KernelDensity devs are happy to add this method, we could just do this, otherwise we would define our own:

```julia
density(; kwargs...) = (args...) -> kde(args...; kwargs...)
density(args...; kwargs...) = kde(args...; kwargs...)
```

The "callable type" can also be made to work but I feel it's quite clumsy in the case where there are no keywords as you'd have to do:

```julia
plot(KDE(), iris, Style(:SepalWidth))
```

That being said, I can quite easily add support to callable types, by adding a method:

```julia
convert_arguments(P::PlotFunc, t::AbstractCallable, args...; kwargs...) =
    convert_arguments(P, (v...) -> t(v...), args...; kwargs...)
```

Is there any strong preference between curried functions and callable structs?

**Comment by piever** [comment]
> Could the `Theme` constructor not simply wrap an input `Vector` in a `Palette` without the user having to worry about this?

I'm actually not sure, it may be better to ask @SimonDanisch, but I'm afraid this would cause issues if users actually want to pass a vector as a keyword argument (say `scatter(rand(3), marker = [:circle, :rect, :cross])` where you want the vector to associate to each dot its respective marker: if the `Theme` constructor happens to be used on this pipeline you'd get a `Palette` instead).

**Pull request merged**